File: Lab_8/Lab_8_java/src/Czytelnia.java
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class Czytelnia {

    private int liczba_czyt = 0;
    private int liczba_pis = 0;
    private int piszacych = 0;
    private Lock lock ;
    private Condition condC;
    private Condition condP;

    public void init()
    {
        lock = new ReentrantLock();
        condC = lock.newCondition();
        condP = lock.newCondition();
    }

    public void chce_czytac() throws InterruptedException
    {
        lock.lock();

        try {
            while (liczba_pis > 0 /*Tutaj brakuje czegoś*/) {
                condC.await();
            }

            liczba_czyt++;
            condC.signal();
        } finally {
            lock.unlock();
        }
    }

    public void czytam()
    {
        System.out.println("Liczba czytających " + liczba_czyt + ", piszących: " + liczba_pis);

        if (liczba_pis > 1 || (liczba_pis == 1 && liczba_czyt > 0) || liczba_pis < 0 || liczba_czyt < 0) {
            System.out.println("Warunek czytania niespełniony");
            System.exit(0);
        }

        try {
            Thread.sleep(300000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public void koniec_czytania()
    {
        lock.lock();

        try {
            liczba_czyt--;


            if (liczba_czyt == 0) {
                lock.notifyAll();
            }
        } finally {
            lock.unlock();
        }
    }

    public void chce_pisac() throws InterruptedException
    {
        lock.lock();

        try {
            if (liczba_czyt + liczba_pis > 0) {
                condP.await();
            }

            liczba_pis++;
        } finally {
            lock.unlock();
        }
    }

    public void pisze()
    {
        System.out.println("Liczba czytających " + liczba_czyt + ", piszących: " + liczba_pis);

        if (liczba_pis>1 || (liczba_pis==1 && liczba_czyt>0) || liczba_pis<0 || liczba_czyt<0 ) {
            System.out.println("Warunek pisania niespełniony");
            System.exit(0);
        }

        try {
            Thread.sleep(300000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public void koniec_pisania() throws InterruptedException
    {
        lock.lock();

        try {
            liczba_pis--;

            if (condC != null) {
                condC.await();
            } else {
                condP.await();
            }
        } finally {
            lock.unlock();
        }
    }

}

File: Lab_8/Lab_8_pthreads/czytelnia.c
#include<stdlib.h>
#include<stdio.h>
#include<unistd.h>
#include<pthread.h>

#include"czytelnia.h"


/*** Implementacja procedur interfejsu ***/

int my_read_lock_lock(czytelnia_t* czytelnia_p){
    pthread_mutex_lock(&czytelnia_p->mutex);

    if(czytelnia_p->liczba_pis > 0 || (czytelnia_p->o_p >0))
    {
    	czytelnia_p->o_c++;
        pthread_cond_wait(&czytelnia_p->cond_c, &czytelnia_p->mutex);
        czytelnia_p->o_c--;
    }
       
    czytelnia_p->liczba_czyt++;
    pthread_cond_signal(&czytelnia_p->cond_c);
    pthread_mutex_unlock(&czytelnia_p->mutex);
}


int my_read_lock_unlock(czytelnia_t* czytelnia_p){
    pthread_mutex_lock(&czytelnia_p->mutex);
    czytelnia_p->liczba_czyt--;

    if(czytelnia_p->liczba_czyt == 0)
        pthread_cond_signal(&czytelnia_p->cond_p);

    pthread_mutex_unlock(&czytelnia_p->mutex);
}


int my_write_lock_lock(czytelnia_t* czytelnia_p){
    pthread_mutex_lock(&czytelnia_p->mutex);

    if((czytelnia_p->liczba_czyt+czytelnia_p->liczba_pis) > 0)
    {
    	czytelnia_p->o_p++;
        pthread_cond_wait(&czytelnia_p->cond_p, &czytelnia_p->mutex);
        czytelnia_p->o_p--;
    }
    czytelnia_p->liczba_pis++;
    pthread_mutex_unlock(&czytelnia_p->mutex);
}


int my_write_lock_unlock(czytelnia_t* czytelnia_p){
    pthread_mutex_lock(&czytelnia_p->mutex);
    czytelnia_p->liczba_pis--;

    if(czytelnia_p->o_c > 0)
        pthread_cond_signal(&czytelnia_p->cond_c);
    else
        pthread_cond_signal(&czytelnia_p->cond_p);

    pthread_mutex_unlock(&czytelnia_p->mutex);
}

void inicjuj(czytelnia_t* czytelnia_p){
    czytelnia_p->liczba_czyt = 0;
    czytelnia_p->liczba_pis = 0;

    pthread_cond_init(&czytelnia_p->cond_c, NULL);
    pthread_cond_init(&czytelnia_p->cond_p, NULL);

    pthread_mutex_init(&czytelnia_p->mutex, NULL);
}

void czytam(czytelnia_t* czytelnia_p){

// wypisanie wartości zmiennych kontrolujących działanie: liczby czytelników i pisarzy
    printf("Liczba czytających %d, piszących: %d\n", czytelnia_p->liczba_czyt, czytelnia_p->liczba_pis);

// sprawdzenie warunku poprawności i ewentualny exit
    if( czytelnia_p->liczba_pis>1 || (czytelnia_p->liczba_pis==1 && czytelnia_p->liczba_czyt>0) || czytelnia_p->liczba_pis<0 || czytelnia_p->liczba_czyt<0 ) 
    { 
        printf("Warunek czytania niespełniony\n"); exit(0); 
    }

    usleep(rand()%300000);
}

void pisze(czytelnia_t* czytelnia_p){

// wypisanie wartości zmiennych kontrolujących działanie: liczby czytelników i pisarzy
    printf("Liczba czytających %d, piszących: %d\n", czytelnia_p->liczba_czyt, czytelnia_p->liczba_pis);


// sprawdzenie warunku poprawności i ewentualny exit
    if( czytelnia_p->liczba_pis>1 || (czytelnia_p->liczba_pis==1 && czytelnia_p->liczba_czyt>0) || czytelnia_p->liczba_pis<0 || czytelnia_p->liczba_czyt<0 ) 
    { 
        printf("Warunek pisania niespełniony\n"); exit(0); 
    }

    usleep(rand()%300000);
}


